File: rust/2021/day_02/Main.java
import java.util.*;
import java.io.*;
import java.util.stream.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new FileReader("input.txt"));
        var splits = br.lines().collect(Collectors.toList());
        long h = 0;
        long d = 0;
        long aim = 0;

        for (int i = 0; i < splits.size(); i++) {
            var curr = splits.get(i).split(" ");
            var direction = curr[0];
            var amount = Integer.valueOf(curr[1]);

            if (direction.equals("forward")) {
                h += amount;
                d += aim * amount;
            } else if (direction.equals("up")) {
                aim -= amount;
            } else if (direction.equals("down")) {
                aim += amount;
            }
        }

        System.out.println(h*d);
    }
}

File: rust/2021/day_02/main.py
with open("input.txt", "r") as file:
    steps = [row.rstrip() for row in file]
    # Part 2
    pos_x = 0
    pos_y = 0
    aim = 0
    for step in steps:
        direction, magnitude = step.split()
        magnitude = int(magnitude)
        if direction == "forward":
            pos_x += magnitude
            pos_y += aim * magnitude
        elif direction == "up":
            aim -= magnitude
        elif direction == "down":
            aim += magnitude
        else:
            print(f"{direction} is an unknown direction!")

    print(pos_x * pos_y)
